### PR TITLE
Depth packing and double sided shadows

### PR DESCRIPTION
**Pull request by tschw**

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/packing.glsl [hunk 1]
```diff
@@ -6,17 +6,26 @@ vec3 unpackRGBToNormal( const in vec3 rgb ) {
...
+const float UnpackDownscale = 255. / 256.; // 0..1 -> fraction (excluding 1)
+
+const vec3 PackFactors = vec3( 256. * 256. * 256., 256. * 256.,  256. );
+const vec4 UnpackFactors = UnpackDownscale / vec4( PackFactors, 1. );
+
+const float ShiftRight8 = 1. / 256.;
+
+vec4 packDepthToRGBA( const in float v ) {
+
+	vec4 r = vec4( fract( v * PackFactors ), v );
```
Comment (bhouston) [inline src/renderers/shaders/ShaderChunk/packing.glsl]: See here for my attempt to make this optimization (fract versus mod):  https://github.com/mrdoob/three.js/pull/8547/files/272056d01bc6325de7d8c306a174b47e634fb918..8dfb3f1d2a53d389540bb514d5fc183605e9b133#r58754365

This refactoring of packing shouldn't really change speed, const floats inside of the function or outside of the function are still pre-computed by the compiler I think.

Comment (bhouston) [inline src/renderers/shaders/ShaderChunk/packing.glsl]: Did you refactor change the resulting constants used?  It is a little bit hard to follow the math.  If you did change them, can you explain in the main PR.

Comment (tschw) [inline src/renderers/shaders/ShaderChunk/packing.glsl]: It's not really a refactor - more of a clean rewrite.

>  It is a little bit hard to follow the math. If you did change them, can you explain in the main PR.
At it...

**Comment by tschw** [comment]
The constants for decoding have changed, because when a float 0..1 is encoded in a 8-bit color channel of a texture, it's not just about putting the bits in there, because the 1 has to fit into the range as well. So values are upscaled and downscaled to make this happen (this is what `PackUpscale` / `UnpackDownscale` compensate). The decoder did not consider this fact, therefore got the lowbits of the bytes wrong.

**Comment by bhouston** [comment]
Interesting.  I wish we had unit tests for the glsl functions as we have had a number of errors in them.  And it would cool to ensure that they do not creep in over time.

I wonder how to do that.  Maybe have one example with a bunch of shaders on the screen that render green if correct and render red if they are wrong?  It could test the decoding/encoding of depth, normals, and the various HDR color spaces (RGBE, RGBD, RGBM, sRGB, LogUV, etc.), and I am sure there are other mathematically functions that we are not testing.  I think edge cases of various BRDF math functions, like total reflection and total absorption, etc would be awesome as well.

I think looking at these periodically and making them right every once in a while isn't a long term solution...

File: src/renderers/webgl/WebGLShadowMap.js [hunk 2]
```diff
@@ -92,6 +92,8 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
 	this.type = THREE.PCFShadowMap;
 	this.cullFace = THREE.CullFaceFront;
 
+	this.allowDoubleSided = false;
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLShadowMap.js]: How about `.forceDoubleSided`?

Comment (mrdoob) [inline src/renderers/webgl/WebGLShadowMap.js]: Actually, ignore that...

File: docs/api/renderers/WebGLRenderer.html [hunk 3]
```diff
@@ -91,27 +91,28 @@
...
 		<div>Defines shadow map type (unfiltered, percentage close filtering, percentage close filtering with bilinear filtering in shader)</div>
 		<div>Options are THREE.BasicShadowMap, THREE.PCFShadowMap, THREE.PCFSoftShadowMap. Default is THREE.PCFShadowMap.</div>
 
 
-		<h3>[property:Integer shadowMapCullFace]</h3>
+		<h3>[property:Integer shadowMap.cullFace]</h3>
 
 		<div>Default is THREE.CullFaceFront. The faces that needed to be culled. Possible values: THREE.CullFaceFront and THREE.CullFaceBack</div>
 
+		<h3>[property:Integer shadowMap.allowDoubleSided]</h3>
```
Comment (mrdoob) [inline docs/api/renderers/WebGLRenderer.html]: Boolean

**Comment by tschw** [comment]
@bhouston 
> See here for my attempt to make this optimization (fract versus mod)
> I discussed it here: #8547 (comment)

I know the history and that `mod` instead of `fract` apparently fixed some precision issues (see @mrdoob 's comment in your review).

However, it wasn't really about the `mod` but the `/ 255`! It made sure the color values are in `[0,1]` instead of `[0,1[` (now in `PackUpscale`).

It wasn't considered in the decoder, so the gaps were gone but values could still overlap. The base-256 digit tuples `(0,255)` and `(1,0)` (big endian notation) decoded to the same value.

> This refactoring of packing shouldn't really change speed, 

Well... Most importantly it changes the correctness from *wrong* to *right* and the clarity from *wild* to *tidy* ;).

It may or may not compile to fewer instructions, depending on how smart the compiler is. 

Since `mod( a, b )` = `fract( a * b ) / b`, going `mod( a * c, b ) / c` may constant-fold `b * c`, but worrying about that we're pretty much missing the point: The original version is still overly complicated code with anonymous constants hard-coded into it. Therefore it should be tidied up before even thinking about optimization.

GLSL compilers do not rearrange arithmetic operations in general. Multiplying a vector with a scalar and with another scalar, it's often exactly what you get, for example. Operation ordering may very well limit constant folding to a certain extent, e.g. I wouldn't expect all implementations to compile `constant * var / constant` to `var`. Also note that I am using 3-vectors where appropriate (to avoid extra instructions with scalar-based GPU - vector-based ones can still pad it up).

> const floats inside of the function or outside of the function are still pre-computed by the compiler I think.

Yes, they are outside the functions because they wanted to live next to each other, I guess :).

**Comment by bhouston** [comment]
This is very cool stuff. :)  Do you know @tschw if ```vec4( x0, x1, x2, x3 ) / y``` has the same efficiency as ```x0 / y```?  Are the four divides done in parallel if you use a vector operation?  Same with operators like ```exp( vec4( x0, x1, x2, x3 ) )``` compared to ```exp( x0 )``` or ```sin( x0 )```.  I guess I most care about the divide operation because I heard those are just brutal on GPUs, pretty much the slowest operation there is.

File: docs/api/renderers/WebGLRenderer.html (hunk 3 above)
Comment (mrdoob) [inline docs/api/renderers/WebGLRenderer.html]: I'll clean up.

**Comment by mrdoob** [comment]
Many thanks!

**Comment by tschw** [comment]
> Do you know @tschw if vec4( x0, x1, x2, x3 ) / y has the same efficiency as x0 / y? Are the four divides done in parallel if you use a vector operation? Same with operators like exp( vec4( x0, x1, x2, x3 ) ) compared to exp( x0 ) or sin( x0 ). I guess I most care about the divide operation because I heard those are just brutal on GPUs, pretty much the slowest operation there is.

This is a super interesting question. Actually several. I don't really have definitive answers to all of them, however I could gather some clues, such as trends in hardware design, information from the GPGPU area, other stuff I read, and educated guesses.

##### Division #####

...in shading languages is multiplication with the reciprocal or something reasonably cheap, so it's just a (probably special) instruction plus maybe a multiplication. Shading languages are specified in ways that allow this kind of implementation.

In GPGPU context, an IEEE 754-compliant division can be a monster. There are intrinsics to get to the cheaper low-level variants. Those may not be that precise and/or incomplete in respect to the IEEE standard.

In case the hardware *can* actually has a division instruction, it is still likely to be a more complex operation than multiplication (NVIDIA GPUs have fewer, custom pipelines for mathematical special functions - and it may very well reside there, for instance). So *I guess* an explicit multiplication with the reciprocal gets you on the safer side in terms of speed in case of `vector / scalar`, but then again it won't be unreasonably expensive so you may as well decide otherwise and opt for more precision.

##### Does vectorized GLSL still matter? #####

Yes, but probably only for mobile platforms. Drivers that have an optimizing shader compiler vectorize our code automagically where need be. They also do a better job than we'd ever want to be doing, since they don't have to care about keeping the code well-structured for human understanding. When you read the second part of the dragon book (I assume at least on guy on a compiler team actually did so) it isn't too hard to implement either.

##### How instructions of the same thread run in parallel - or not #####

Modern, compute-capable GPUs map scalar instructions to an array of parallel [pipelines](https://en.wikipedia.org/wiki/Instruction_pipelining). SIMD operations in the GLSL may still execute in parallel, as may independent scalar ones, or execute time-displaced, or may even be torn apart and execute sequentially:

NVIDIA was the first GPU vendor to move away from vectors *within* the threads ~7 years ago. AMD moved away from them with [Graphics Core Next (see "Compute Units" 2nd paragraph)](https://en.wikipedia.org/wiki/Graphics_Core_Next#Compute_Units), HD7700+ ~ 4 years ago, and Intel recently moved away from the approach as well. 

Vectorization then only happens on scalar registers across groups of threads in lock-step execution (called "warps" in CUDA or "wavefronts" in OpenCL), pretty much similar to "ray packet traversal", which means using SIMD registers with scalar-based code to trace "packets" of multiple rays at once, but more automated. NVIDIA coined the term SIMT - "single instruction multiple threads" for this scheme, but it has been around in supercomputing long before that. It's sometimes also confusingly called SIMD (which it still is, just applied differently). Similarly confusing may be to find the terms "vector" and "scalar" in [AMD's GCN ISA specs](http://amd-dev.wpengine.netdna-cdn.com/wordpress/media/2013/07/AMD_GCN3_Instruction_Set_Architecture.pdf) to refer to "a scalar per thread" (a vector of 64 values) and "one single scalar for the entire wavefront", respectively; the assembly program, other than the GLSL one, describes the wavefront (that is how to shade 64 fragments at once in case of a fragment shader, for example) - not a single thread therein.

The "SIMT thing" is not only done for more flexible scheduling, which matters for compute, but also for energy efficiency - vector hardware can't always fully be utilized, yet the silicon needs to be powered:

In each core, the scheduler switches through the currently active warps (or wavefronts) and issues instructions to a shared array of parallel pipelines. This way, the pipelines fill up nicely, and there's enough time for the instructions to complete in a time-displaced, parallel fashion before another warpful of instructions that need the results may be issued, as long as there is enough work to be done. Note that this approach allows to decouple control flow divergence to a certain extent (another story for another time - forcing branchless code is actually only a good idea when there are just few conditional instructions involved, otherwise branching code will be more efficient). 

Actually, multiple instructions from one warp can go into the pipelines upon one visit of the scheduler, so the pipelines still fill nicely with fewer work (e.g. a vertex shader invocation on only a few vertices, or other warps aren't ready due to synchronization).
These instructions must be independent or data hazards must be resolved (AFAIK GPUs don't do the latter, because the multiplexing would either become way too heavy or become limiting on how instructions can be mapped to the pipelines). The hardware figures out the [ILP](https://en.wikipedia.org/wiki/Instruction-level_parallelism) at runtime. This [SO post](http://stackoverflow.com/questions/9886002/how-does-the-cuda-warp-scheduler-issue-2-instructions-at-a-time-for-a-warp) discusses the NVIDIA case. Pretty much poor man's [out-of-order execution](https://en.wikipedia.org/wiki/Out-of-order_execution), first documented [here](http://www.cs.berkeley.edu/~volkov/volkov10-GTC.pdf). I guess the others have followed, by now.

File: docs/api/renderers/WebGLRenderer.html (hunk 3 above)
Comment (tschw) [inline docs/api/renderers/WebGLRenderer.html]: Just noticed you did. Thanks :)
